clippy_dev/src/crater.rs

```diff
@@ -0,0 +1,182 @@
+            let ungz_tar = flate2::read::GzDecoder::new(std::fs::File::open(&krate_file_path).unwrap());
+            // extract the tar archive
+            let mut archiv = tar::Archive::new(ungz_tar);
+            archiv.unpack(&extract_dir).expect("Failed to extract!");
+        }
+        // crate is extracted, return a new Krate object which contains the path to the extracted
+        // sources that clippy can check
+        Krate {
+            version: self.version.clone(),
+            name: self.name.clone(),
+            path: extract_dir.join(format!("{}-{}/", self.name, self.version)),
+        }
+    }
+}
+
+impl Krate {
+    fn run_clippy_lints(&self, cargo_clippy_path: &PathBuf) -> Vec<String> {
+        println!("Linting {} {}...", &self.name, &self.version);
+        let cargo_clippy_path = std::fs::canonicalize(cargo_clippy_path).unwrap();
+
+        let shared_target_dir = clippy_project_root().join("target/crater/shared_target_dir/");
+
+        let all_output = std::process::Command::new(cargo_clippy_path)
+            .env("CARGO_TARGET_DIR", shared_target_dir)
+            // lint warnings will look like this:
+            // src/cargo/ops/cargo_compile.rs:127:35: warning: usage of `FromIterator::from_iter`
+            .args(&[
+                "--",
+                "--message-format=short",
+                "--",
+                "--cap-lints=warn",
+                "-Wclippy::pedantic",
+                "-Wclippy::cargo",
+            ])
+            .current_dir(&self.path)
+            .output()
+            .unwrap();
+        let stderr = String::from_utf8_lossy(&all_output.stderr);
+        let output_lines = stderr.lines();
+        let mut output: Vec<String> = output_lines
+            .into_iter()
+            .filter(|line| line.contains(": warning: "))
+            // prefix with the crate name and version
+            // cargo-0.49.0/src/cargo/ops/cargo_compile.rs:127:35: warning: usage of `FromIterator::from_iter`
+            .map(|line| format!("{}-{}/{}", self.name, self.version, line))
+            // remove the "warning: "
+            .map(|line| {
+                let remove_pat = "warning: ";
+                let pos = line
+                    .find(&remove_pat)
+                    .expect("clippy output did not contain \"warning: \"");
+                let mut new = line[0..pos].to_string();
+                new.push_str(&line[pos + remove_pat.len()..]);
+                new.push('\n');
+                new
+            })
+            .collect();
+
+        // sort messages alphabtically to avoid noise in the logs
+        output.sort();
+        output
+    }
+}
+
+fn build_clippy() {
+    Command::new("cargo")
+        .arg("build")
+        .output()
+        .expect("Failed to build clippy!");
+}
+
+// the main fn
+pub fn run() {
+    let cargo_clippy_path: PathBuf = PathBuf::from("target/debug/cargo-clippy");
+
+    // crates we want to check:
+    let krates: Vec<KrateSource> = vec![
```

Comment:
As far as I understand,
```toml
[crates]
crate1 = ".."
crate2= ".."
```
would mean that we have a hashmap of crates...?
I wonder if this is an unnecessary limit because in theory, there is nothing that would keep us from checking version 1 of a crate but also version 2 of the same crate.

Reply:
Hm good point. Maybe also accept a list for the version numbers then.

Reply:
We can now list several crate versions: https://github.com/rust-lang/rust-clippy/pull/6469/files#diff-ab04b04dd70029c1a47f1122e55d10b705efb593b7da586039f3cd19e03e99df